// @cc-web/server/dist/jsonl.js
const NOISE_TYPES = new Set([
    'system',
    'mode',
    'permission-mode',
    'file-history-snapshot',
    'last-prompt',
    'attachment',
]);
export function parseJsonl(content) {
    const lines = content.split('\n').filter(line => line.trim());
    const messages = [];
    for (const line of lines) {
        try {
            const parsed = JSON.parse(line);
            // Skip noise types
            if (!parsed.type || NOISE_TYPES.has(parsed.type)) {
                continue;
            }
            // Handle user messages: type="user" with message.content
            if (parsed.type === 'user' && parsed.message?.content) {
                messages.push({
                    role: 'user',
                    content: typeof parsed.message.content === 'string'
                        ? parsed.message.content
                        : JSON.stringify(parsed.message.content),
                    timestamp: parseTimestamp(parsed.timestamp),
                    model: parsed.message.model,
                });
                continue;
            }
            // Handle assistant messages: type="assistant" with message.content array
            if (parsed.type === 'assistant' && parsed.message?.content) {
                const contentArray = Array.isArray(parsed.message.content)
                    ? parsed.message.content
                    : [parsed.message.content];
                // Extract text from content array
                const textParts = [];
                for (const item of contentArray) {
                    if (typeof item === 'object' && item.type === 'text' && item.text) {
                        textParts.push(item.text);
                    }
                }
                if (textParts.length > 0) {
                    messages.push({
                        role: 'assistant',
                        content: textParts.join('\n\n'),
                        timestamp: parseTimestamp(parsed.timestamp),
                        model: parsed.message.model,
                    });
                }
                continue;
            }
            // Handle legacy text format: type="text" with text field
            if (parsed.type === 'text' && parsed.text) {
                messages.push({
                    role: 'assistant',
                    content: parsed.text,
                    timestamp: parseTimestamp(parsed.timestamp),
                    model: parsed.model,
                });
                continue;
            }
        }
        catch {
            // Skip malformed lines
            continue;
        }
    }
    return messages;
}
function parseTimestamp(ts) {
    if (!ts)
        return Date.now();
    if (typeof ts === 'number')
        return ts;
    return new Date(ts).getTime();
}
